airflow/models/dagrun.py

```diff
@@ -207,10 +207,26 @@ def next_dagruns_to_examine(
                 DagModel.is_paused == expression.false(),
                 DagModel.is_active == expression.true(),
             )
-            .order_by(
-                nulls_first(cls.last_scheduling_decision, session=session),
-                cls.execution_date,
+        )
+        if state == State.QUEUED:
+            stmt = (
+                session.query(DagModel, func.count(DagModel.dag_id))
+                .filter(DagModel.is_active == expression.true(), DagModel.is_paused == expression.false())
+                .join(DagRun, DagRun.dag_id == DagModel.dag_id)
+                .filter(DagRun.state == State.RUNNING)
+                .group_by(DagModel.dag_id)
+                .all()
             )
+            dag_ids = []
+            for dag_model, total_runs in stmt:
+                if total_runs >= dag_model.max_active_runs:
+                    dag_ids.append(dag_model.dag_id)
+
+            query = query.filter(~cls.dag_id.in_(dag_ids))
```

Comment:
This will work for small number of DAGs, but for large numbers (1000s or tens of 1000s) this is going to put extra load on the DB by counting the running dag runs ever time, even if we don't care about that dag

```python
            running_drs = (
                session.query(DagRun.dag_id, func.count(DagRun.state).label('num_running'))
                .filter(DagRun.state == DagRunState.RUNNING)
                .group_by(DagRun.dag_id)
                .cte('running_dag_runs')
            )

            query = query.outerjoin(running_drs, running_drs.c.dag_id == DagRun.dag_id).filter(
                func.coalesce(running_drs.c.num_running, 0) < DagModel.max_active_runs
            )
```

Reply:
Wow! This worked like a charm! Thank you!!